code/code/misc/multiclass.cc

```diff
@@ -150,42 +150,18 @@ int TBeing::getClassNum(classIndT arg) const
 }
 
 
-classIndT TBeing::getClassIndNum(const char *arg, exactTypeT exact) const
-{
-  int which = getClassNum(arg, exact);
-  classIndT res = MIN_CLASS_IND;
-
-  for(classIndT i=MIN_CLASS_IND;i<MAX_CLASSES;i++){
-    if(classInfo[i].class_num==which){
-      res=i;
-      break;
-    }
-  }
-      
-  if(res==MIN_CLASS_IND) {
-    vlogf(LOG_BUG, "unknown class result");
-  }
+classIndT TBeing::getClassIndNum(const char *arg, exactTypeT exact) const {
+  auto which = static_cast<unsigned short>(getClassNum(arg, exact));
 
-  return res;
+  return getClassIndNum(which);
 }
 
+classIndT TBeing::getClassIndNum(unsigned short which) const {
+  for (classIndT i = MIN_CLASS_IND; i < MAX_CLASSES; i++)
+    if (classInfo[i].class_num == which)
+      return i;
```

Comment:
Regarding a lookup table, isn't that essentially what `classInfo` is or am I misunderstanding what you mean by that?

```c++
const struct class_info classInfo[MAX_CLASSES] =
{
  {true, MAGE_LEVEL_IND, CLASS_MAGE, "mage", 
   DISC_MAGE, DISC_LORE, 0.43, 7.5 * 35.0 / 50.0, "M"},

  {true, CLERIC_LEVEL_IND, CLASS_CLERIC, "cleric",
   DISC_CLERIC, DISC_THEOLOGY, 0.47, 8.0 * 35.0 / 50.0, "C"},

  {true, WARRIOR_LEVEL_IND, CLASS_WARRIOR, "warrior",
   DISC_WARRIOR, DISC_NONE, 0.47, 8.5, "W"},

  {true, THIEF_LEVEL_IND, CLASS_THIEF, "thief",
   DISC_THIEF, DISC_NONE, 0.41, 8.0 * 35.0 / 50.0, "T"},

  {true, SHAMAN_LEVEL_IND, CLASS_SHAMAN, "shaman",
   DISC_SHAMAN, DISC_NONE, 0.39, 7.5 * 35.0 / 50.0, "S"},

  {true, DEIKHAN_LEVEL_IND, CLASS_DEIKHAN, "deikhan",
   DISC_DEIKHAN, DISC_THEOLOGY, 0.44, 7.5, "D"},

  {true, MONK_LEVEL_IND, CLASS_MONK, "monk",
   DISC_MONK, DISC_NONE, 0.44, 7.5 * 35.0 / 50.0, "K"},

  {false, RANGER_LEVEL_IND, CLASS_RANGER, "ranger",
   DISC_RANGER, DISC_NONE, 0.46, 7.0 * 35.0 / 50.0, "R"},

  {false, COMMONER_LEVEL_IND, CLASS_COMMONER, "commoner",
   DISC_ADVENTURING, DISC_NONE, 0.40, 5, "O"}
};
```

Except it's indexed by the `classIndT` enum, and we get an `unsigned short` representing a class bitflag passed in as our only identifying info.

I could create a `map` to translate an incoming class bitflag into a  `classIndT` - is that what you mean by lookup table?

```c++
static const std::unordered_map<uint16_t, classIndT> CLASS_BITFLAG_TO_CLASS_IND_T = {
    { CLASS_MAGE, MAGE_LEVEL_IND},
    // { ... and so on }
};

classIndT TBeing::getClassIndNum(unsigned short which) const {
    if (which < CLASS_MAGE || which > CLASS_RANGER)
        return MAX_CLASSES;
    return CLASS_BITFLAG_TO_CLASS_IND_T.at(which);
};
```

Guess it just depends whether that's worth it to avoid one for loop? Unless there's an easier way to do it than this, which I wouldn't doubt.

Reply:
Yes, I meant about the same thing as `classInfo` but indexed by `which`.
Also, `map` is not the only option - you could use a regular array, or a vector with `at`, populated with `class_num`s at the correct array positions - if the number of elements is small, then it's faster to iterate over a small part of cached-once memory than walk a tree structure and incurring a bunch of cache misses. Hash maps might be faster, too.

Not that anybody cares about the speed of this.

Reply:
With your blessing I think I'm going to just leave this as it is, in order to get this PR done with. Everything works great at this point, and I'll come back and replace this loop with a lookup table in the future if profiling suggests it would be worth it. Or as part of a general cleanup/modernization pass on the entire file.